src/object/srv_obj.c

```diff
@@ -2342,19 +2328,115 @@ ds_obj_tgt_update_handler(crt_rpc_t *rpc)
 	obj_ioc_end(&ioc, rc);
 }
 
+/* Extract local iods/offs/csums by orw_oid.id_shard from @orw */
+static int
+obj_tgt_update_split(struct obj_rw_in *orw, struct daos_oclass_attr *oca,
+		     daos_iod_t **iods, uint64_t **offs,
+		     struct dcs_iod_csums **csums, struct obj_pool_metrics *opm)
+{
+	struct obj_shard_iod	*siod;
+	uint32_t		local_tgt;
+	uint32_t		oiod_nr;
+	int			i;
+	int			rc = 0;
+
+	oiod_nr = orw->orw_iod_array.oia_oiod_nr;
+	D_ASSERT(oiod_nr > 0);
+	if (oiod_nr > 1) {
+		D_ALLOC_ARRAY(*iods, oiod_nr);
+		if (*iods == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		D_ALLOC_ARRAY(*offs, oiod_nr);
+		if (*offs == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		if (csums != NULL) {
+			D_ALLOC_ARRAY(*csums, oiod_nr);
+			if (*csums == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			for (i = 0; i < oiod_nr; i++) {
+				D_ALLOC_PTR((*csums)[i].ic_data);
+				if ((*csums)[i].ic_data == NULL)
+					D_GOTO(out, rc = -DER_NOMEM);
+			}
+		}
+	}
+
+	local_tgt = orw->orw_oid.id_shard % obj_ec_tgt_nr(oca);
+	for (i = 0; i < oiod_nr; i++) {
+		daos_iod_t		*iod_parent;
+		struct dcs_iod_csums	*iod_pcsum = NULL;
+		struct obj_io_desc	*oiod;
+
+		iod_parent = &orw->orw_iod_array.oia_iods[i];
+		oiod = &orw->orw_iod_array.oia_oiods[i];
+
+		if (opm != NULL)
+			obj_ec_metrics_process(iod_parent, oiod, oca, opm);
+
+		memcpy(&(*iods)[i], iod_parent, sizeof(daos_iod_t));
+		if (csums != NULL) {
+			iod_pcsum = &orw->orw_iod_array.oia_iod_csums[i];
+			memcpy(&(*csums)[i].ic_akey, &iod_pcsum->ic_akey,
+			       sizeof(struct dcs_csum_info));
+			(*csums)[i].ic_nr = iod_pcsum->ic_nr;
+		}
+
+		if ((*iods)[i].iod_type == DAOS_IOD_ARRAY) {
+			siod = obj_shard_iod_get(oiod, local_tgt);
+			D_ASSERTF(siod != NULL, "local_tgt %u\n", local_tgt);
+			(*offs)[i] = siod->siod_off;
+			(*iods)[i].iod_recxs = &iod_parent->iod_recxs[siod->siod_idx];
+			(*iods)[i].iod_nr = siod->siod_nr;
+			if (csums != NULL) {
+				(*csums)[i].ic_data = &iod_pcsum->ic_data[siod->siod_idx];
```

Comment:
Should be "*(*csums)[i].ic_data = iod_pcsum->ic_data[siod->siod_idx];" ?

Reply:
oh, we can re-use parent RPC ic_data here, so only assigning the pointer here would be enough, IMHO.

Reply:
But you have already alloc space for "(*csums)[i].ic_data" at L2358.

Reply:
yes, indeed. we need alloc space for single value case, though can re-use the pointer for array value.

Besides, for most of cases, iod_nr will be 1, i.e. there should not be any allocation normally, which is also the patch for.

Reply:
Some confused, I mean that if we have already allocated space for "(*csums)[i].ic_data", then here, we re-assign it with new pointer, then we leak DRAM, right?